### ISSUE TITLE
[compiled autograd][cppnode] eliminate recompiles on ctx->saved_data

### ISSUE THREAD

**Issue by xmfan** [pytorch/pytorch#130170]
### 🚀 The feature, motivation and pitch

User-defined autograd functions may use ctx->saved_data to pass non-Tensor activations e.g. int, float, bool, etc. These are passed as [IValues](https://github.com/pytorch/pytorch/blob/18e75c098b219f2d5103e93c60f1de91fd06fa3e/aten/src/ATen/core/ivalue.h#L221), and currently we specialize on them, resulting in recompiles any time there's any value change.

Workaround that exist today:
- move all ctx->saved_data scalar conversions into a custom op to hide it from the compiler
- rewrite the c++ autograd function to only use tensors
- (if you're just trying to get perf numbers while this is unsupported) build from source and comment out https://github.com/pytorch/pytorch/blob/18e75c098b219f2d5103e93c60f1de91fd06fa3e/torch/csrc/autograd/custom_function.h#L205

Supporting dynamism for non-Tensors requires a bit of work because:
- The user-defined backward will generally call conversion methods e.g. [FBGEMM PermutePooledEmbsFunction](https://github.com/pytorch/FBGEMM/blob/eb73980029c39ed7dbd4bf963fc28bb7f5012590/fbgemm_gpu/src/permute_pooled_embedding_ops/permute_pooled_embedding_function.cpp#L70-L71) `ctx->saved_data["allow_duplicates"].toBool();`, which will return a new object https://github.com/pytorch/pytorch/blob/3d56673b24073235eb3beff3f8cf1a88a91ba459/aten/src/ATen/core/ivalue.h#L1341-L1354 This is annoying for tracing a dynamic graph, since we want all scalars to use their symbolic equivalents but the API returns a new object that cannot be swapped by compiled autograd
- Dynamo has a few different codepaths to specialize on scalar values, but these lifted scalars from saved activations frequently change and specialization here will cause us to recompile

Proposed fixes:
1. Derived class that makes IValue underlying scalars swappable and override conversion methods to return symbolic variables tracked by compiled autograd
2. Dynamo + dynamic support for keeping SymFloat/SymBool/(Sym?)String as lifted inputs

### Alternatives

A possible, but annoying alternative, is to ask users to rewrite their autograd function to only use Tensors, and move their Tensor to scalar conversion in custom ops

cc @ezyang @anijain2305 @chauhang @penguinwu

**Comment by zou3519** [comment]
Does changing calls to .toInt() into .toSymInt() work for the int case? If so I'd expect we can do something analogous for bools and floats we want to be symbolic

**Comment by xmfan** [comment]
Yes, changing the user code calls to use .toSymInt will work, and it removes the need to edit IValue internals

**Comment by xmfan** [comment]
@zou3519 Does the custom op schema support Symint as input arguments?

I'm trying to support an op redispatch where one of the input is some int/float/bool we passed from forward. Not sure how to convert from a symbolic c10::SymInt (the fake one created by compiled autograd) back into an actual int. In python, we can use the node's hint, but in c++ the node (SymNodeImpl?) doesn't have hints implemented

**Comment by zou3519** [comment]
> Does the custom op schema support Symint as input arguments?

Yes

**Comment by xmfan** [comment]
So we have a lack of support for SymFloat, while I did land https://github.com/pytorch/pytorch/pull/133048 for the compiled autograd side:
1. backend may not support SymFloat as mentioned in this comment: https://github.com/pytorch/pytorch/blob/d4b31f7bcf962560768b1ee24a2ee5a1e914b8ba/torch/_dynamo/config.py#L58-L61 I tried with some toy examples in inductor, and I got some FakeTensor backend errors
2. I don't think custom op schema supports SymFloat. See the test case: https://github.com/pytorch/pytorch/pull/133278 and its gdb backtrace https://gist.github.com/xmfan/f6a77d5b6faa8d739657f39f6e6988d9

**Comment by zou3519** [comment]
@xmfan why do you need SymFloat? We haven't seen a situation where this is necessary

**Comment by xmfan** [comment]
@rzou Say the user saves a learning rate (that changes every iteration) for backward, compiled autograd needs to lift it using IValue::toSymFloat to avoid recompiling every time.

Need to turn off specialize_float: Once compiled autograd traced a graph with a float as input, dynamo will specialize on the float input and recompile every time the float changes.

Need in custom op schema: If the user wanted to use that learning rate in some op/kernel, the signature needs to accept SymFloat for compiled autograd's proxy to associate the op call with the lifted float.

**Comment by zou3519** [comment]
The workaround here is to put the learning rate into a Tensor. I don't know if better support for SymFloat is going to come anytime soon (@ezyang ?)

**Comment by ezyang** [comment]
urghhhhh I'm actually done with the design and poc but there's a bunch of just grinding that needs to be done that I haven't had time in the last month to do

**Comment by xmfan** [comment]
this is indirectly addressed with the initial capture rewrite, we always graph break on cppnodes lol
